### DISCUSSION TITLE
Configuration of prebuild ipxe image

### DISCUSSION THREAD

**Discussion by pide2000** [ipxe/ipxe#614] [Q&A] [answered]
Hi Community, 

I try to build an ipxe iso image on my own. Therefore I want to "simply" build an ipxe iso image which is basically the same as the prebuild one except that I want to enable https. 
My Problem is, when I build from scratch and define DOWNLOAD_PROTO_HTTPS, I get an iso image which is not bootable. By comparing the prebuid iso and my own, I found out that the file esp.img is missing in my iso.
I guess I missunderstood something or I have to define something to get it work. 
Maybe someone can provide me the config and build steps for the prebuild image so I can start learning from this.

Thanks 
Peter

**Comment by stappersg** [comment]
Maybe first build an unmodified version?

In case that already has been done, tell us, _the part of the iPXE community that follows github.com discussions_, how that went.

**Comment by NiKiZe** [comment]
How do you build the image, what do you boot on (efi/pcbios) and how?

Could it be that you don't generate the current files to later build the iso? I would avoid the iso as a first step and use other images in the first place, or at least as a step to verify the build.

**Comment by pide2000** [comment]
Thanks for the answers.
@stappersg yes I've done this. Image is build without errors but it doesn't boot. 

@NiKiZe 
Its an efi and I simply dd the iso image to an sd card. This works for the prebuild out of the box. Because I want to play around with https I considered to enable https and build an image on my own. This fails. 

> Could it be that you don't generate the current files to later build the iso? I would avoid the iso as a first step and use other images in the first place, or at least as a step to verify the build.

I followed the instructions sayin `cd /src && make`, maybe I was a little naive. :)  
And maybe my assumption, prebuild iso is the outcome of `cd /src && make`, is completely missleading. 

It will help me to understand if someone can tell me the steps to reproduce the prebuild image, I think, or at least I hope I can gain more insights. Learning by doing. But of course, any support is appreciated.

**Reply by mcb30** [reply]
> I followed the instructions sayin `cd /src && make`, maybe I was a little naive. :) And maybe my assumption, prebuild iso is the outcome of `cd /src && make`, is completely missleading.

The default build will give you a BIOS image.  To build an EFI image, you will need to do something like `make bin-x86_64-efi/ipxe.iso`.

For the record: the prebuilt ISO published at https://boot.ipxe.org/ipxe.iso is a combination of the BIOS, x86_64 EFI, and Aarch64 EFI images, built using the equivalent of:
```
make bin/ipxe.lkrn bin-x86_64-efi/ipxe.efi
make bin-arm64-efi/ipxe.efi CROSS=aarch64-linux-gnu-
./util/genfsimg -o ipxe.iso bin/ipxe.lkrn bin-x86_64-efi/ipxe.efi bin-arm64-efi/ipxe.efi
```
The Makefile does not currently support building this combined ISO in one step as a single target, since it involves compiling for multiple CPU architectures.

**Reply by NiKiZe** [reply]
For your usecase, build ipxe.efi, have the USB formatted with fat32, copy ipxe.efi to \efi\boot\bootx64.efi

**Reply by stappersg** [reply]
And when it still doesn't boot, elaborate your _but it doesn't boot_. Use http://www.catb.org/~esr/faqs/smart-questions.html as guidance for how. Earn karma bonus points by sharing your succes.

**Reply by pide2000** [reply]
Hey Guys, 

thank you very much for your support. 
It works for me with the following steps:´
```
make bin/ipxe.lkrn bin-x86_64-efi/ipxe.efi
./util/genfsimg -o ipxe.iso bin/ipxe.lkrn bin-x86_64-efi/ipxe.efi 

```
I then changed config/general.h and config/console.h for my needs and everything works perfectly. 

 
@NiKiZe I will try your suggestion in the near future. 
@stappersg plz send a little karma :)  I hope this thread helps others to get started. 

What about arm? 
I opted out arm because I get some errors at compiling and I don't need it for now. Furthermore I think its better to start a new discussion in case I can't figure it out on my own.

**Comment by pide2000** [comment]
I just released an article on medium.com about iPXE. 
https://medium.com/@peter.bolch/how-to-netboot-with-ipxe-6a41db514dee
Maybe it's interesting for someone an maybe I get some more Karma ;)

**Reply by stappersg** [reply]
Regarding _karma_, two screenshots for the _iPXE IRC channel_.
Point of view is my, `stappers`,  IRC client programm in timezone _CET_ (and _CEST_)
The blurly parts are on purpose.

![image](https://user-images.githubusercontent.com/35434094/160284778-a5100685-956a-450a-afa2-a3928b1bb061.png)

![image](https://user-images.githubusercontent.com/35434094/160284866-22447b65-f6ae-49c6-9fca-e814c99ea2f9.png)

**Comment by pide2000** [comment, marked as answer]
Copy the reply of @mcb30 so I can mark this as answer. 

> The default build will give you a BIOS image. To build an EFI image, you will need to do something like make bin-x86_64-efi/ipxe.iso.
> 
> For the record: the prebuilt ISO published at https://boot.ipxe.org/ipxe.iso is a combination of the BIOS, x86_64 EFI, and Aarch64 EFI images, built using the equivalent of:
> 
> make bin/ipxe.lkrn bin-x86_64-efi/ipxe.efi
> make bin-arm64-efi/ipxe.efi CROSS=aarch64-linux-gnu-
> ./util/genfsimg -o ipxe.iso bin/ipxe.lkrn bin-x86_64-efi/ipxe.efi bin-arm64-efi/ipxe.efi
> 
> The Makefile does not currently support building this combined ISO in one step as a single target, since it involves compiling for multiple CPU architectures.
